internal/controllers/addon/monitoring_stack_reconciler.go

```diff
@@ -0,0 +1,225 @@
+	desiredMonitoringStack, err := r.getDesiredMonitoringStack(ctx, addon)
+	if err != nil {
+		return nil, err
+	}
+
+	// returns observed MonitoringStack object
+	reconciledMonitoringStack, err := r.reconcileMonitoringStack(ctx, desiredMonitoringStack)
+	if err != nil {
+		return nil, err
+	}
+	return reconciledMonitoringStack, nil
+}
+
+func (r *monitoringStackReconciler) propagateMonitoringStackStatusToAddon(monitoringStack *obov1alpha1.MonitoringStack, addon *addonsv1alpha1.Addon) (monitoringStackStackAvailable bool) {
+	availableCondition, reconciledCondition := obov1alpha1.Condition{}, obov1alpha1.Condition{}
+	availableConditionFound, reconciledConditionFound := false, false
+
+	for _, cond := range monitoringStack.Status.Conditions {
+		cond := cond
+		if cond.Type == obov1alpha1.AvailableCondition {
+			availableCondition = cond
+			availableConditionFound = true
+		} else if cond.Type == obov1alpha1.ReconciledCondition {
+			reconciledCondition = cond
+			reconciledConditionFound = true
+		}
+		if availableConditionFound && reconciledConditionFound {
+			break
+		}
+	}
+
+	if availableConditionFound && availableCondition.Status == obov1alpha1.ConditionTrue {
+		return true
+	}
+
+	if availableConditionFound && availableCondition.Status != obov1alpha1.ConditionTrue {
+		reportUnreadyMonitoringStack(addon, fmt.Sprintf("MonitoringStack Unavailable: %s", availableCondition.Message))
+	} else if reconciledConditionFound {
+		if reconciledCondition.Status == obov1alpha1.ConditionTrue {
+			reportUnreadyMonitoringStack(addon, "MonitoringStack successfully reconciled: Pending MonitoringStack to be Available")
+		} else {
+			reportUnreadyMonitoringStack(addon, fmt.Sprintf("MonitoringStack failed to reconcile: %s", reconciledCondition.Message))
+		}
+	} else {
+		reportUnreadyMonitoringStack(addon, "MonitoringStack pending to get reconciled")
+	}
+
+	return false
+}
+
+// helper function to generate desired MonitoringStack object
+func (r *monitoringStackReconciler) getDesiredMonitoringStack(ctx context.Context,
+	addon *addonsv1alpha1.Addon) (*obov1alpha1.MonitoringStack, error) {
+
+	commonConfig, stop := parseAddonInstallConfig(controllers.LoggerFromContext(ctx), addon)
+	if stop {
+		return nil, fmt.Errorf("error parsing Addon config")
+	}
+
+	var (
+		remoteWriteURL     string
+		oauthConfig        *monv1.OAuth2
+		writeRelabelConfig []monv1.RelabelConfig
+	)
+
+	rhobsRemoteWriteConfig := addon.Spec.Monitoring.MonitoringStack.RHOBSRemoteWriteConfig.DeepCopy()
+	if rhobsRemoteWriteConfig != nil {
+		remoteWriteURL = rhobsRemoteWriteConfig.URL
+		oauthConfig = rhobsRemoteWriteConfig.OAuth2.DeepCopy()
+		writeRelabelConfig = getWriteRelabelConfigFromAllowlist(rhobsRemoteWriteConfig.Allowlist)
+	}
+
+	desiredMonitoringStack := &obov1alpha1.MonitoringStack{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getMonitoringStackName(addon.Name),
+			Namespace: commonConfig.Namespace,
+		},
+		Spec: obov1alpha1.MonitoringStackSpec{
```

Comment:
Just curious, how would the obo operator know which prometheus instances to visit to get the metrics listed in 
`Allowlist`

Reply:
That would be communicated via the ServiceMonitors/PodMonitors which will, in turn, be developed and lifecycled by the addon. Just one pre-requisite, those ServiceMonitors/PodMonitors must have the label `addons.managed.openshift.io/mso: <addon-name>` because the Prometheus instances spun up MonitoringStack CR would inherit scrape_configs from the ServiceMonitor/PodMonitor with the label `addons.managed.openshift.io/mso: <addon-name>`. 

Ref: we configure `addons.managed.openshift.io/mso: <addon-name>` to be the resourceSelector under the moniitoringStack CR created by ADO.